### ISSUE TITLE
E731 Doesn't Ignore Dataclass Attributes

### ISSUE THREAD

**Issue by max-muoto** [astral-sh/ruff#13336]
While `E731` correctly flags class attributes that are assigned to lambdas:

```python
class container:
    foo = lambda x: x
```

In the case of dataclases it incorrectly flags the assignment:

```python
@dataclasses.dataclass
class Container:
    foo = lambda x: x
```

From my POV this should be functionally equivalent to setting a default in an `__init__`, which is allowed. Anything with [`dataclass_transform`](https://peps.python.org/pep-0681/) applied (Pydantic models, msgspec structs, etc) should likely additionally be ignored, however this would require more advanced introspection.

**Comment by dhruvmanila** [comment]
Related to https://github.com/astral-sh/ruff/issues/10718, I think if you annotate the variable Ruff should ignore this: https://play.ruff.rs/5b6b4a77-94eb-4245-bc14-6f3d6bc61f22

**Comment by charliermarsh** [comment]
Annotating it as a callable changes the meaning, though. It becomes an assignable field in the constructor. I still think it's fine to flag this -- you _can_ assign it in the `__init__`. And if we don't flag this, why would we flag the non-dataclass case?
